twoliter/src/lock.rs

```diff
@@ -139,6 +195,121 @@ impl Lock {
+            .cloned()
+            .context(format!(
+                "could not find kit image for architecture '{}' at {}",
+                docker_arch, image.source
+            ))
+    }
+
+    async fn get_index<P>(&self, path: P, digest_uri: &str) -> Result<IndexView>
+    where
+        P: AsRef<Path>,
+    {
+        let oci_archive = docker!(
+            ["save", digest_uri],
+            format!("failed to fetch kit and save to disk from {}", digest_uri)
+        );
+        let mut oci_archive = Archive::new(oci_archive.as_slice());
+        // Unpack the oci image to tempdirectory
+        oci_archive
+            .unpack(path.as_ref())
+            .context("failed to unpack oci image")?;
+        // Now read in the index
+        let index_bytes = read(path.as_ref().join("index.json")).await?;
+        serde_json::from_slice(index_bytes.as_slice())
+            .context("failed to deserialize oci image index")
+    }
+
+    async fn extract_kit<P>(&self, path: P, image: &LockedImage, arch: &str) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let tmp_dir =
+            tempdir().context("failed to create a temporary directory for kit extraction")?;
+        let vendor = image.vendor.clone();
+        let name = image.name.clone();
+        let docker_arch = match arch {
+            "x86_64" => Ok("amd64"),
+            "aarch64" => Ok("arm64"),
+            _ => Err(anyhow!("unsupported architecture '{}'", arch)),
+        }?;
+        let target_path = path.as_ref().join(format!("{vendor}/{name}/{arch}"));
+        create_dir_all(&target_path).await?;
+
+        // First we want to get the correct digest for our architecture
+        let manifest = self.get_manifest(image, docker_arch).await?;
+        let image_source = image.digest_uri(manifest.digest.as_str());
+        let index = self
+            .get_index(tmp_dir.path(), image_source.as_str())
+            .await?;
+        let digest = index
+            .manifests
+            .first()
+            .context("empty oci image")?
+            .digest
+            .replace(':', "/");
+        let manifest_bytes = read(tmp_dir.path().join(format!("blobs/{digest}")))
+            .await
+            .context("failed to read manifest blob")?;
+        let manifest_layout: ManifestLayoutView = serde_json::from_slice(manifest_bytes.as_slice())
+            .context("failed to deserialize oci manifest")?;
+        for layer in manifest_layout.layers {
+            let digest = layer.digest.replace(':', "/");
+            let mut layer_blob = File::open(tmp_dir.path().join(format!("blobs/{digest}")))
+                .context("failed to read layer of oci image")?;
+            let decompressed = flate2::read::GzDecoder::new(&mut layer_blob);
+            let mut layer_archive = Archive::new(decompressed);
+            layer_archive
+                .unpack(&target_path)
+                .context("failed to unpack layer to disk")?;
```

Comment:
It's not clear where / when we might be retrieving large files or loading them into memory.

Where does the primary content fetch happen? Not the metadata but the potentially GiBs of RPMs. It must be `get_manifest` or `get_index` but those functions sound like they should be fast, and do not imply "do a giant download as a side effect".

Reply:
The primary content fetch occurs in get_index as that is when we need the oci image to use data with docker save. I'll rename the function